derive/src/lib.rs

```diff
@@ -245,9 +245,21 @@ fn generate_variant_type(data_enum: &DataEnum, scale_info: &Ident) -> TokenStrea
         .filter(|v| !utils::should_skip(&v.attrs))
         .map(|v| {
             let ident = &v.ident;
-            let v_name = quote! {stringify!(#ident) };
-            match v.fields {
-                Fields::Named(ref fs) => {
+            let v_name = quote! { stringify!(#ident) };
+            let index = utils::maybe_index(v);
+            match (&v.fields, index) {
+                (Fields::Named(ref fs), Some(index)) => {
+                    let fields = generate_fields(&fs.named);
+                    quote! {
+                        .indexed_variant(
+                            #v_name,
+                            #index,
+                            :: #scale_info::build::Fields::named()
+                                #( #fields)*
+                        )
+                    }
+                }
+                (Fields::Named(ref fs), None) => {
```

Comment:
This whole block would be simplified if `.variant(` accepted an `index: Option<u8>`, instead of having two methods. What do you think?

Reply:
I go back and forth on that. This code here sure is ugly; otoh the API itself is straight forward: there's no way users can misunderstand what `indexed_variant()` does.

The best option would be a builder, i.e. to have:

```
Variants::with_fields()
    .variant(
        "A", 
        Fields::unnamed()
            .field_of::<u8>("u8")
    )
    .indexed(123)
```

…but I think that's fiddly to get working.

Reply:
In other words: if you have a strong opinion, I'll follow your lead! ;)

Reply:
Given that we have another optional field on the way (docs) #87, maybe it is worth now introducing a variant builder. Hopefully that would also simplify this big block of code.

Reply:
Builder added and merged  from #87 into this branch